File: src/mymcp/transfer/endpoints.py
"""Bypass HTTP routes for binary file transfer.

These endpoints authenticate via the URL ticket alone — no Bearer header.
Bytes flow as raw streams; nothing about file content enters the MCP
protocol or LLM context.
"""

from __future__ import annotations

import contextlib
import logging
import os
import tempfile

from fastapi import FastAPI, Request
from starlette.responses import JSONResponse, StreamingResponse

from mymcp import config
from mymcp.audit import log_tool_call
from mymcp.tools.files import check_protected_path
from mymcp.transfer import get_ticket_store

logger = logging.getLogger("mymcp")


class _SizeExceeded(Exception):
    pass


def _err(status: int, code: str, hint: str) -> JSONResponse:
    return JSONResponse({"ok": False, "error": code, "hint": hint}, status_code=status)


def _disabled_response() -> JSONResponse:
    return _err(404, "transfer_disabled", "File transfer is disabled on this server.")


def _client_ip(request: Request) -> str:
    return request.client.host if request.client else "unknown"


def _content_disposition(filename: str) -> str:
    """Build a safe Content-Disposition value.

    Filenames may contain ``"``, ``\\`` or control chars (``\\r``/``\\n``)
    which would break header parsing or enable header injection. Strip
    control chars, escape ``\\`` and ``"`` for the quoted ``filename=``,
    and add an RFC 5987 ``filename*=UTF-8''…`` for non-ASCII names.
    """
    from urllib.parse import quote

    safe_ascii = "".join(c for c in filename if ord(c) >= 0x20 and c != "\x7f")
    quoted = safe_ascii.replace("\\", "\\\\").replace('"', '\\"')
    ascii_only = quoted.encode("ascii", "ignore").decode("ascii")
    star = quote(safe_ascii, safe="")
    return f"attachment; filename=\"{ascii_only}\"; filename*=UTF-8''{star}"


def _audit_redeem(
    ticket,
    *,
    success: bool,
    bytes_count: int,
    error_code: str | None,
    client_ip: str,
) -> None:
    log_tool_call(
        token_name=ticket.created_by,
        role="rw" if ticket.op == "upload" else "ro",
        ip=client_ip,
        tool="transfer_redeem",
        params={
            "op": ticket.op,
            "path": ticket.path,
            "ticket": ticket.ticket_id[:8],
            "bytes": bytes_count,
        },
        result="ok" if success else "error",
        error_code=error_code,
        error_message=error_code if not success else None,
    )


def register_transfer_routes(app: FastAPI) -> None:
    @app.put("/files/raw/{ticket_id}")
    async def upload_endpoint(ticket_id: str, request: Request):
        if not config.TRANSFER_ENABLED:
            return _disabled_response()
        store = get_ticket_store()
        ticket = store.lookup(ticket_id)
        if ticket is None:
            kind = store.classify(ticket_id)
            if kind == "expired":
                return _err(410, "ticket_expired", "Mint a new ticket.")
            if kind == "consumed":
                return _err(410, "ticket_not_found", "Ticket already used.")
            return _err(404, "ticket_not_found", "Mint a new ticket.")
        if ticket.op != "upload":
            return _err(405, "wrong_method", "This ticket requires GET.")
        if not store.consume(ticket_id):
            return _err(410, "ticket_not_found", "Ticket already used.")
        return await _do_upload(ticket, request)

    @app.get("/files/raw/{ticket_id}")
    async def download_endpoint(ticket_id: str, request: Request):
        if not config.TRANSFER_ENABLED:
            return _disabled_response()
        store = get_ticket_store()
        ticket = store.lookup(ticket_id)
        if ticket is None:
            kind = store.classify(ticket_id)
            if kind == "expired":
                return _err(410, "ticket_expired", "Mint a new ticket.")
            if kind == "consumed":
                return _err(410, "ticket_not_found", "Ticket already used.")
            return _err(404, "ticket_not_found", "Mint a new ticket.")
        if ticket.op != "download":
            return _err(405, "wrong_method", "This ticket requires PUT.")
        if not store.consume(ticket_id):
            return _err(410, "ticket_not_found", "Ticket already used.")
        return await _do_download(ticket, request)


async def _do_upload(ticket, request: Request):
    ip = _client_ip(request)
    err = check_protected_path(ticket.path)
    if err:
        _audit_redeem(
            ticket, success=False, bytes_count=0, error_code="path_protected", client_ip=ip
        )
        return _err(403, "path_protected", err)

    declared = request.headers.get("content-length")
    if declared is not None:
        try:
            if int(declared) > ticket.max_bytes:
                _audit_redeem(
                    ticket,
                    success=False,
                    bytes_count=int(declared),
                    error_code="size_exceeded",
                    client_ip=ip,
                )
                return _err(
                    413,
                    "size_exceeded",
                    f"Body exceeds max_bytes={ticket.max_bytes}.",
                )
        except ValueError:
            _audit_redeem(
                ticket, success=False, bytes_count=0, error_code="bad_content_length", client_ip=ip
            )
            return _err(400, "bad_content_length", "Content-Length is not an integer.")

    parent = os.path.dirname(ticket.path) or "/"
    try:
        os.makedirs(parent, exist_ok=True)
    except OSError as e:
        _audit_redeem(ticket, success=False, bytes_count=0, error_code="mkdir_failed", client_ip=ip)
        return _err(500, "mkdir_failed", str(e))

    fd, tmp_path = tempfile.mkstemp(prefix=".mymcp-upload-", dir=parent)
    written = 0
    try:
        with os.fdopen(fd, "wb") as out:
            async for chunk in request.stream():
                if not chunk:
                    continue
                if written + len(chunk) > ticket.max_bytes:
                    raise _SizeExceeded()
                out.write(chunk)
                written += len(chunk)
        os.replace(tmp_path, ticket.path)
    except _SizeExceeded:
        with contextlib.suppress(OSError):
            os.unlink(tmp_path)
        _audit_redeem(
            ticket, success=False, bytes_count=written, error_code="size_exceeded", client_ip=ip
        )
        return _err(413, "size_exceeded", f"Body exceeds max_bytes={ticket.max_bytes}.")
    except Exception as e:  # pragma: no cover - defensive
        with contextlib.suppress(OSError):
            os.unlink(tmp_path)
        logger.error("upload failed for %s: %s", ticket.path, e)
        _audit_redeem(
            ticket, success=False, bytes_count=written, error_code="write_failed", client_ip=ip
        )
        return _err(500, "write_failed", str(e))

    _audit_redeem(ticket, success=True, bytes_count=written, error_code=None, client_ip=ip)
    return JSONResponse({"ok": True, "path": ticket.path, "bytes_written": written})


async def _do_download(ticket, request: Request):
    ip = _client_ip(request)
    err = check_protected_path(ticket.path)
    if err:
        _audit_redeem(
            ticket, success=False, bytes_count=0, error_code="path_protected", client_ip=ip
        )
        return _err(403, "path_protected", err)
    if not os.path.isfile(ticket.path):
        _audit_redeem(
            ticket, success=False, bytes_count=0, error_code="path_not_found", client_ip=ip
        )
        return _err(404, "path_not_found", "Server file no longer exists.")

    size = os.path.getsize(ticket.path)
    filename = os.path.basename(ticket.path)
    file_path = ticket.path
    captured_ticket = ticket

    def iter_file():
        sent = 0
        success = False
        error_code: str | None = "stream_aborted"
        try:
            with open(file_path, "rb") as fh:
                while True:
                    chunk = fh.read(64 * 1024)
                    if not chunk:
                        break
                    sent += len(chunk)
                    yield chunk
            success = True
            error_code = None
        finally:
            _audit_redeem(
                captured_ticket,
                success=success,
                bytes_count=sent,
                error_code=error_code,
                client_ip=ip,
            )

    headers = {
        "content-length": str(size),
        "content-disposition": _content_disposition(filename),
    }
    return StreamingResponse(iter_file(), media_type="application/octet-stream", headers=headers)

File: go/internal/transfer/endpoints.go
package transfer

import (
	"encoding/json"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/algony-tony/mymcp/go/internal/audit"
	"github.com/algony-tony/mymcp/go/internal/fsutil"
)

// Endpoints serves the ticket-only /files/raw routes. It shares the TicketStore
// with the prepare_* tools and writes transfer_upload/transfer_download audit.
type Endpoints struct {
	Tickets   *TicketStore
	Audit     *audit.Writer
	Protected []fsutil.ProtectedEntry
	Enabled   bool
	// OnAuditFail is invoked when a redemption audit record cannot be written;
	// it bumps mymcp_audit_write_failures_total (SOC: audit loss must be
	// visible). Nil disables the callback (used by tests).
	OnAuditFail func()
}

// Register wires PUT+GET /files/raw/{ticket_id} onto mux.
func (e *Endpoints) Register(mux *http.ServeMux) {
	mux.HandleFunc("PUT /files/raw/{ticket_id}", e.Upload)
	mux.HandleFunc("GET /files/raw/{ticket_id}", e.Download)
}

func writeErr(w http.ResponseWriter, status int, code, hint string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(map[string]any{"ok": false, "error": code, "hint": hint})
}

func clientIP(r *http.Request) string {
	if host, _, err := net.SplitHostPort(r.RemoteAddr); err == nil {
		return host
	}
	return "unknown"
}

// resolveTicket runs the shared lookup/classify/method/consume gate. Returns the
// consumed ticket, or nil after having written the error response.
func (e *Endpoints) resolveTicket(w http.ResponseWriter, r *http.Request, wantOp string) *Ticket {
	if !e.Enabled {
		writeErr(w, 404, "transfer_disabled", "File transfer is disabled on this server.")
		return nil
	}
	id := r.PathValue("ticket_id")
	tk := e.Tickets.Lookup(id)
	if tk == nil {
		switch e.Tickets.Classify(id) {
		case "expired":
			writeErr(w, 410, "ticket_expired", "Mint a new ticket.")
		case "consumed":
			writeErr(w, 410, "ticket_not_found", "Ticket already used.")
		default:
			writeErr(w, 404, "ticket_not_found", "Mint a new ticket.")
		}
		return nil
	}
	if tk.Op != wantOp {
		if wantOp == "upload" {
			writeErr(w, 405, "wrong_method", "This ticket requires GET.")
		} else {
			writeErr(w, 405, "wrong_method", "This ticket requires PUT.")
		}
		return nil
	}
	if !e.Tickets.Consume(id) {
		writeErr(w, 410, "ticket_not_found", "Ticket already used.")
		return nil
	}
	return tk
}

// auditRedeem writes one transfer_upload/transfer_download record. On write
// failure it bumps mymcp_audit_write_failures_total (via OnAuditFail) and
// returns the error so a confirmable mutation is not acknowledged unaudited —
// silent audit loss is a SOC red line.
func (e *Endpoints) auditRedeem(tk *Ticket, ok bool, n int64, code, ip string) error {
	tool := "transfer_download"
	if tk.Op == "upload" {
		tool = "transfer_upload"
	}
	result := "ok"
	var errCode, errMsg string
	if !ok {
		result, errCode, errMsg = "error", code, code
	}
	err := e.Audit.Log(audit.Entry{
		TS: time.Now().UTC().Format(time.RFC3339Nano), TokenName: tk.CreatedBy,
		Role: tk.CreatedByRole, IP: ip, Tool: tool, Result: result,
		ErrorCode: errCode, ErrorMessage: errMsg,
		Params: map[string]any{
			"op": tk.Op, "path": tk.Path, "ticket": firstN(tk.TicketID, 8),
			"bytes": n, "issuer_token_name": tk.CreatedBy,
			"issuer_role": tk.CreatedByRole, "redeemer_ip": ip,
		},
	})
	if err != nil && e.OnAuditFail != nil {
		e.OnAuditFail()
	}
	return err
}

func firstN(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// Upload handles PUT /files/raw/{ticket_id}.
func (e *Endpoints) Upload(w http.ResponseWriter, r *http.Request) {
	tk := e.resolveTicket(w, r, "upload")
	if tk == nil {
		return
	}
	ip := clientIP(r)
	if msg := fsutil.CheckProtectedPath(tk.Path, fsutil.ModeWrite, e.Protected); msg != "" {
		e.auditRedeem(tk, false, 0, "path_protected", ip)
		writeErr(w, 403, "path_protected", msg)
		return
	}
	if cl := r.Header.Get("Content-Length"); cl != "" {
		if n, err := strconv.ParseInt(cl, 10, 64); err != nil {
			e.auditRedeem(tk, false, 0, "bad_content_length", ip)
			writeErr(w, 400, "bad_content_length", "Content-Length is not an integer.")
			return
		} else if n > tk.MaxBytes {
			e.auditRedeem(tk, false, n, "size_exceeded", ip)
			writeErr(w, 413, "size_exceeded", "Body exceeds max_bytes="+strconv.FormatInt(tk.MaxBytes, 10)+".")
			return
		}
	}
	parent := filepath.Dir(tk.Path)
	if parent == "" {
		parent = "/"
	}
	if err := os.MkdirAll(parent, 0o755); err != nil {
		e.auditRedeem(tk, false, 0, "mkdir_failed", ip)
		writeErr(w, 500, "mkdir_failed", err.Error())
		return
	}
	tmp, err := os.CreateTemp(parent, ".mymcp-upload-*")
	if err != nil {
		e.auditRedeem(tk, false, 0, "write_failed", ip)
		writeErr(w, 500, "write_failed", err.Error())
		return
	}
	tmpPath := tmp.Name()
	written, exceeded, copyErr := copyCapped(tmp, r.Body, tk.MaxBytes)
	tmp.Close()
	if exceeded {
		os.Remove(tmpPath)
		e.auditRedeem(tk, false, written, "size_exceeded", ip)
		writeErr(w, 413, "size_exceeded", "Body exceeds max_bytes="+strconv.FormatInt(tk.MaxBytes, 10)+".")
		return
	}
	if copyErr != nil {
		os.Remove(tmpPath)
		e.auditRedeem(tk, false, written, "write_failed", ip)
		writeErr(w, 500, "write_failed", copyErr.Error())
		return
	}
	if err := os.Rename(tmpPath, tk.Path); err != nil {
		os.Remove(tmpPath)
		e.auditRedeem(tk, false, written, "write_failed", ip)
		writeErr(w, 500, "write_failed", err.Error())
		return
	}
	if err := e.auditRedeem(tk, true, written, "", ip); err != nil {
		// The bytes are on disk, but do not confirm an unauditable mutation.
		writeErr(w, 500, "audit_failed", "upload succeeded but the audit record could not be written")
		return
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]any{"ok": true, "path": tk.Path, "bytes_written": written})
}

// copyCapped streams src→dst, stopping if more than limit bytes arrive.
func copyCapped(dst io.Writer, src io.Reader, limit int64) (written int64, exceeded bool, err error) {
	buf := make([]byte, 64*1024)
	for {
		n, rerr := src.Read(buf)
		if n > 0 {
			if written+int64(n) > limit {
				return written, true, nil
			}
			if _, werr := dst.Write(buf[:n]); werr != nil {
				return written, false, werr
			}
			written += int64(n)
		}
		if rerr == io.EOF {
			return written, false, nil
		}
		if rerr != nil {
			return written, false, rerr
		}
	}
}

// Download handles GET /files/raw/{ticket_id}.
func (e *Endpoints) Download(w http.ResponseWriter, r *http.Request) {
	tk := e.resolveTicket(w, r, "download")
	if tk == nil {
		return
	}
	ip := clientIP(r)
	if msg := fsutil.CheckProtectedPath(tk.Path, fsutil.ModeRead, e.Protected); msg != "" {
		e.auditRedeem(tk, false, 0, "path_protected", ip)
		writeErr(w, 403, "path_protected", msg)
		return
	}
	st, err := os.Stat(tk.Path)
	if err != nil || st.IsDir() {
		e.auditRedeem(tk, false, 0, "path_not_found", ip)
		writeErr(w, 404, "path_not_found", "Server file no longer exists.")
		return
	}
	f, err := os.Open(tk.Path)
	if err != nil {
		e.auditRedeem(tk, false, 0, "path_not_found", ip)
		writeErr(w, 404, "path_not_found", "Server file no longer exists.")
		return
	}
	defer f.Close()
	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Length", strconv.FormatInt(st.Size(), 10))
	w.Header().Set("Content-Disposition", contentDisposition(filepath.Base(tk.Path)))
	sent, cerr := io.Copy(w, f)
	if cerr != nil {
		e.auditRedeem(tk, false, sent, "stream_aborted", ip)
		return
	}
	e.auditRedeem(tk, true, sent, "", ip)
}

// contentDisposition builds a safe attachment header (port of _content_disposition).
func contentDisposition(filename string) string {
	var b strings.Builder
	for _, c := range filename {
		if c >= 0x20 && c != 0x7f {
			b.WriteRune(c)
		}
	}
	safe := b.String()
	quoted := strings.ReplaceAll(safe, `\`, `\\`)
	quoted = strings.ReplaceAll(quoted, `"`, `\"`)
	asciiOnly := strings.Map(func(r rune) rune {
		if r > 127 {
			return -1
		}
		return r
	}, quoted)
	star := url.PathEscape(safe)
	return `attachment; filename="` + asciiOnly + `"; filename*=UTF-8''` + star
}
